### DISCUSSION TITLE
Is there some sort of a firewall on Spirallinux?

### DISCUSSION THREAD

**Discussion by Xispra** [SpiralLinux/SpiralLinux-project#15] [Q&A] [answered]
I installed ssh and tried to connect to my SL install, but the connection was refused. So was an effort to ping the IP address. When I went downstairs to my Spiral computer (not a box 7et, I still haven't put the components in a case) I couldn't even ping localhost as root! And I can't ssh even from localhost. I've never run into this before.

I looked around systemsettings5 & the menus but didn't see anything. ufw doesn't seem to be installed, that seems to be one of the more popular Linux firewalls.

I don't have much experience with firewalls, I haven't really used them since I stopped using Windows in the mid aughts. But I looked at Discover (which I'm not familiar with, I almost always use apt/apt-get, and synaptic when necessary), but didn't find anything. So I did a:

```
apt search firewall | grep installed

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

  installed tests for firewalld
libnftables1/stable,now 0.9.8-3.1 amd64 [installed]
nftables/stable,now 0.9.8-3.1 amd64 [installed]
```

So nftables looks like some sort of firewall/packet filter. I don't want to break anything so I did a: 
_systemctl stop  nftables.service_

That seemed to kill the service.
```
systemctl status  nftables.service
● nftables.service - nftables
     Loaded: loaded (/lib/systemd/system/nftables.service; disabled; vendor preset: e>
     Active: inactive (dead)
       Docs: man:nft(8)
             http://wiki.nftables.org
lines 1-5/5 (END)
```
Nonetheless I still can't ping localhost. 

Is nftables what is causing this behavior? 

Do I need to disable or uninstall the service and reboot?  Will that muck anything up? I'm not running a net-facing server, just a local desktop machine.

Is there some GUI interface to control nftables that I overlooked? 

If nftables is the issue, and inadvisable to remove, is there some web page you would recommend to get started? I need to enable ping, ssh & samba to my local network, and possibly some other stuff.

I am befuddled.

**Comment by geckolinux** [comment, marked as answer]
Hi there, thanks for giving SpiralLinux a try. Sorry for the difficulty with what you're trying to accomplish.

I don't explicitly include a firewall in SpiralLinux because it doesn't run internet-facing services by default and because most users will already be behind a NAT firewall. However, I used [this](https://github.com/nodiscc/debian-live-config) project as a template to learn the Debian live-build system, which includes some custom tweaks to some low-level settings. Since I'm not as familiar with Debian's quirks I assumed that they were probably there for a reason and I left some of them in place. I suspect that your problem lies in `/etc/sysctl.d/custom` , the quickest way to confirm would probably be by moving the `custom` file out of that directory and rebooting.

Please let us know if that worked, and if that was the problem maybe you can help me decide if any or all of those settings should be maintained or removed from the next release of SpiralLinux.

----------------
***EDIT:*** _It's almost surely one or all of these settings in_ `/etc/sysctld.d/custom`  : 

```
# Ignore ICMP redirects
net.ipv4.conf.all.accept_redirects = 0
net.ipv4.conf.all.secure_redirects = 0

# Ignore any ICMP ECHO requests (ping)
net.ipv4.icmp_echo_ignore_all = 1

# Ignore all ICMP ECHO and TIMESTAMP requests received vai broadcast/multicast
net.ipv4.icmp_echo_ignore_broadcasts = 1

**Reply by Xispra** [reply]
Thanks for the hint, geckolinux. I tried moving that file elsewhere, in my case it was /etc/sysctl.d/custom.conf and not  /etc/sysctl.d/custom. After rebooting I could ping in, but still couldn't ssh, I got: 
_ssh: connect to host nuc port 22: Connection refused_

So at least some of the stuff in /etc/sysctl.d/custom.conf looked useful, in particular I liked a line _vm.swappiness = 10_ (# Do not swap unless necessary). I've had systems practically grind to a halt even when they had plenty of RAM left over because of excessive swappiness. So I copied the custom.conf file back to /etc/sysctl.d and just changed the ping line, rebooted, and can now ping but still can't ssh.
```
# Ignore any ICMP ECHO requests (ping)
### net.ipv4.icmp_echo_ignore_all = 1
net.ipv4.icmp_echo_ignore_all = 0
```
So I looked over my /etc/ssh/sshd_config & everything looked OK. But when I tried to start the service I got:
```
# systemctl start sshd.service 
Job for ssh.service failed because the control process exited with error code.
See "systemctl status ssh.service" and "journalctl -xe" for details.
```
Same for _systemctl start ssh_   The journalctl was not too helpful, except maybe one line:
_sshd: no hostkeys available -- exiting._
```
ls -la /etc/ssh/ssh_host_rsa_key
ls: cannot access '/etc/ssh/ssh_host_rsa_key': No such file or directory
```
That is weird, installing ssh has always automatically generated host keys for me automagically. This does not seem to be expected behavior according to the [Debian SSH Wiki](https://wiki.debian.org/SSH).

So I did a:
```
ssh-keygen -A
systemctl start ssh.service
```
Now I have keys and can start the service and ssh to the box from other machine(s).  

Incidentally, I tried to add some verbiage to /etc/nftables.conf  to enable ssh. But apparently the service is not running.
```
# systemctl status nftables
● nftables.service - nftables
     Loaded: loaded (/lib/systemd/system/nftables.service; d>
     Active: failed (Result: exit-code) since Thu 2022-06-16>
       Docs: man:nft(8)
             http://wiki.nftables.org
   Main PID: 5750 (code=exited, status=1/FAILURE)
        CPU: 7ms

Jun 16 12:49:49 nuc systemd[1]: Starting nftables...
Jun 16 12:49:49 nuc nft[5750]: /etc/nftables.conf:18:1-5: Er>
Jun 16 12:49:49 nuc nft[5750]: table inet firewall {
Jun 16 12:49:49 nuc nft[5750]: ^^^^^
Jun 16 12:49:49 nuc systemd[1]: nftables.service: Main proce>
Jun 16 12:49:49 nuc systemd[1]: nftables.service: Failed wit>
Jun 16 12:49:49 nuc systemd[1]: Failed to start nftables.
root@nuc:/etc/ssh# systemctl restart nftables
```

Since I don't have nftables on my (*buntu focal) local server and everything seems to work I just removed my edits to the nftables.conf file and will ignore the fact that the service doesn't seem to work, and maybe remove it at some point. 

So to reiterate, I seem to have had two problems. The /etc/sysctl.d/custom.conf seems to block ping for some reason. And installing openssh-server and/or ssh does not generate keys, unlike any other time I've done that with Debian/*buntu. And although it doesn't seem germane to my issues nftables is installed but apparently not working properly. These are issues to keep an eye on and maybe consider not adding nftables to the default install and modifying the default /etc/sysctl.d/custom.conf file the next go-round. Well, time to reboot, if everything is not working I'll check back in. Regards...

**Reply by Xispra** [reply]
After a reboot pinging & ssh were fine, at least from localhost (which is all I tried), but nftables still seems unhappy.
```
# systemctl status nftables
● nftables.service - nftables
     Loaded: loaded (/lib/systemd/system/nftables.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
       Docs: man:nft(8)
             http://wiki.nftables.org
root@nuc:/home/xispra# systemctl start nftables
Job for nftables.service failed because the control process exited with error code.
See "systemctl status nftables.service" and "journalctl -xe" for details.

```
I saw this:
_/etc/nftables.conf:18:1-1: Error: syntax error, unexpected end of file_
I tried _dpkg-reconfigure nftables_ & purging and reinstalling nftables but it still seemed unhappy.
```
# systemctl status nftables.service
● nftables.service - nftables
     Loaded: error (Reason: Unit nftables.service failed to load properly: Device or resource busy.)
     Active: failed (Result: exit-code) since Thu 2022-06-16 13:40:54 PDT; 2min 26s ago
       Docs: man:nft(8)
             http://wiki.nftables.org
   Main PID: 2601 (code=exited, status=1/FAILURE)
        CPU: 7ms

Jun 16 13:40:54 nuc systemd[1]: Starting nftables...
Jun 16 13:40:54 nuc systemd[1]: nftables.service: Main process exited, code=exited, status=1/FAILURE
Jun 16 13:40:54 nuc nft[2601]: /etc/nftables.conf:18:1-1: Error: syntax error, unexpected end of file
Jun 16 13:40:54 nuc nft[2601]: #
Jun 16 13:40:54 nuc nft[2601]: ^
Jun 16 13:40:54 nuc systemd[1]: nftables.service: Failed with result 'exit-code'.
Jun 16 13:40:54 nuc systemd[1]: Failed to start nftables
```

Seems like purging and reinstalling should have fixed the "unexpected end of file" issue. The /etc/nftables.conf file looks exactly like it did before I tried editing it earlier. I'm not sure what is going on, but doesn't seem to be a big issue in my use-case.

**Reply by geckolinux** [reply]
Excellent, thanks a lot for all the details.

- So, I'll definitely remove the ICMP tweaks from `/etc/sysctl.d/custom.conf` . Thanks for catching and reporting that.
- I agree that the nftables thing is an unrelated issue. It hasn't been explicitly added to the build recipe, but the `nftables` package can also be removed from the installed system without affecting anything else, so I'm honestly not sure how it got installed. But it is disabled by default, so it shouldn't cause any problems, and I confirmed that the service does start normally if manually enabled under a fresh SpiralLinux installation.
- Regarding SSH not starting, I also was running into that during the testing phase and I couldn't figure out why. Thanks a ton for the `ssh-keygen -A` tip, I never would have figured that out. ;-) It appears that is the default intended behavior in Debian ([link 1](https://groups.google.com/g/linux.debian.bugs.dist/c/Q8fabWaVwuY/m/m68-DOf_AwAJ) , [link 2](https://askubuntu.com/a/962393)). I'm pretty sure I can make the Calamares installer run that automatically on the installed system, although I'm not sure if the haveged entropy service will run in the chroot environment to ensure entropy. Or maybe I could rip off that Rasbian service mentioned in [link 2](https://askubuntu.com/a/962393) and make it run during the first system boot. Any suggestions?

Thanks a lot for all of your useful observations to make SpiralLinux better out-of-the-box.

**Reply by geckolinux** [reply]
Little update:
I'm going to completely remove the `/etc/sysctl.d/custom.conf` . There were really a huge number of tweaks in there, many of which I don't understand, so I think it's best to trust that Debian already uses sane defaults.
